File: M3Tools/Controls/Google/Drive/FileUpload.vb
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows.Forms
Imports SPPBC.M3Tools.GTools.Types

Public Class FileUpload

	Public ReadOnly Property Files As IList
		Get
			Return bsFiles.List
		End Get
	End Property

	<DefaultValue(True)>
	Public Property MultiSelect As Boolean
		Get
			Return ofd_FileDialog.Multiselect
		End Get
		Set(value As Boolean)
			ofd_FileDialog.Multiselect = value
		End Set
	End Property

	Public Property DataSource As BindingSource
		Get
			Return bsFiles 'CType(dgv_Files.DataSource, BindingSource)
		End Get
		Set(value As BindingSource)
			'dgv_Files.AutoGenerateColumns = False
			'dgv_Files.DataSource = value
			bsFiles = value
		End Set
	End Property

	Private Sub LoadFiles(sender As Object, e As CancelEventArgs) Handles ofd_FileDialog.FileOk
		For Each file In ofd_FileDialog.FileNames
			If Duplicate(file) Then
				Continue For
			End If
			bsFiles.Add(New File("", file, file.Split("."c)(1)))
		Next
	End Sub

	Private Function Duplicate(fileName As String) As Boolean
		' TODO: Find how to simplify
		For Each file As File In bsFiles.List
			If file.Name = fileName Then
				Return True
			End If
		Next

		Return False
	End Function

	Private Sub SelectFiles(sender As Object, e As ToolStripItemClickedEventArgs) Handles ts_Tools.ItemClicked
		UseWaitCursor = True
		Dim res = ofd_FileDialog.ShowDialog()
		If res = DialogResult.OK Then
			UseWaitCursor = False
		End If
	End Sub

	Private Sub DataSourceUpdated(sender As Object, e As ListChangedEventArgs) Handles bsFiles.ListChanged
		If e.ListChangedType = ListChangedType.ItemAdded OrElse e.ListChangedType = ListChangedType.ItemDeleted Then
			bsFiles.ResetBindings(False)
		End If
	End Sub
End Class

File: M3Tools/Controls/Google/Drive/FileUpload.cs
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

namespace SPPBC.M3Tools
{
	/// <summary>
	/// A control to upload files to Google Drive
	/// </summary>
    public partial class FileUpload
    {
		/// <summary>
		/// The list of files being uploaded
		/// </summary>
        public IList Files
        {
            get
            {
                return bsFiles.List;
            }
        }

		/// <summary>
		/// Whether mulitple files can be uploaded at once
		/// </summary>
        [DefaultValue(true)]
        public bool MultiSelect
        {
            get
            {
                return ofd_FileDialog.Multiselect;
            }
            set
            {
                ofd_FileDialog.Multiselect = value;
            }
        }

		/// <summary>
		/// The datasource to bind to the control
		/// </summary>
        public BindingSource DataSource
        {
            get
            {
                return bsFiles; // CType(dgv_Files.DataSource, BindingSource)
            }
            set
            {
                // dgv_Files.AutoGenerateColumns = False
                // dgv_Files.DataSource = value
                bsFiles = value;
            }
        }

		/// <summary>
		/// 
		/// </summary>
        public FileUpload()
        {
            InitializeComponent();
        }

        private void LoadFiles(object sender, CancelEventArgs e)
        {
            foreach (var @file in ofd_FileDialog.FileNames)
            {
                if (Duplicate(@file))
                {
                    continue;
                }
                bsFiles.Add(new Types.GTools.File("", @file, @file.Split('.')[1]));
            }
        }

        private bool Duplicate(string fileName)
        {
            // TODO: Find how to simplify
            foreach (Types.GTools.File @file in bsFiles.List)
            {
                if ((@file.Name ?? "") == (fileName ?? ""))
                {
                    return true;
                }
            }

            return false;
        }

        private void SelectFiles(object sender, ToolStripItemClickedEventArgs e)
        {
            UseWaitCursor = true;
            var res = ofd_FileDialog.ShowDialog();
            if (res == DialogResult.OK)
            {
                UseWaitCursor = false;
            }
        }

        private void DataSourceUpdated(object sender, ListChangedEventArgs e)
        {
            if (e.ListChangedType == ListChangedType.ItemAdded || e.ListChangedType == ListChangedType.ItemDeleted)
            {
                bsFiles.ResetBindings(false);
            }
        }
    }
}
